---
title: "Computational Life Science Seminar"
author: "Stefan Schmutz"
date: "2020-11-16"
output:
  xaringan::moon_reader:
    lib_dir: libs
    nature:
      titleSlideClass: [center, middle]
      ratio: "16:9"
      slideNumberFormat: ""
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
---




```{r setup, include=FALSE}

knitr::opts_chunk$set(fig.retina = 3)

options(htmltools.dir.version = FALSE)

xaringanExtra::use_xaringan_extra(c("tile_view", "tachyons"))

library(tidyverse)
library(xaringanthemer)

style_mono_accent(
  base_color = "#000000",
  header_font_google = google_font("Cabin"),
  text_font_google   = google_font("Noto Sans")
)
```

background-image: url(figures/title.png)
background-position: center
background-size: 75%


<!-- set style for all following captions --> 
<style>
p.caption {
  font-size: 0.6em;
  font-style: italic;
}
</style>

???

Hi everyone. Please interrupt me anytime if you'd like to ask something.  

Today, I'd like to present this article by Kelleher and colleagues which was published last year and carries the title "Inferring whole-genome hisories in large population data sets"

---
class: center, middle

# What does this **Title** reveal?

???

What does it reveal about the contents of the article?  

As it turns out quite a lot.  

Let's have a look at it bit-by-bit.


---


## .bg-washed-red.b--light-red.ba.bw2.br3.ph2.pv1[Inferring] .moon-gray[.bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[whole-genome] .bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[histories] in .bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[large population datasets]]

<br/>

```{r fig.cap="source: thesaurus.com", echo=FALSE, out.width="500px", fig.align="center"}
knitr::include_graphics("figures/inferring_adjusted2.jpg")
```

???

*Inferring* in this context means that:

The authors apply a model (result might therefore not be 100% accurate, but it's a **good guess/hypothesis**)  
**TODO: check if it's correct to say "apply a model" or if there's a better way to describe this**

---

## .moon-gray[.bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[Inferring]] .bg-washed-yellow.b--yellow.ba.bw2.br3.ph2.pv1[whole-genome] .moon-gray[.bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[histories] in .bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[large population datasets]]

```{r fig.cap="source: National Human Genome Research Institute", echo=FALSE, out.width="400px", fig.align="center"}
knitr::include_graphics("figures/Karyotype.png")
```

???

Next, *whole-genome*  

Here they work with whole-genome data from humans.  
The human genome (blueprint for an individual) consists of 23 chromosome pairs (it's diploid), shown in this figure, which are made of DNA.  
DNA in turn consists of four bases (A, T, G, C) and, it's order of those bases (sequence) can be determined using DNA sequencing methods.  

Since we're a sexually reproducing species, we carry parts of the genome from both of our parents in us.  



---

## .moon-gray[.bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[Inferring] .bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[whole-genome]] .bg-washed-green.b--dark-green.ba.bw2.br3.ph2.pv1[histories] .moon-gray[in .bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[large population datasets]]

???

Over time the DNA sequence can change (evolutionary process).  
Observing these changes of the DNA allows us to reconstruct history (genealogy).  
A way to visualize this history is through phylogenetic trees.  

--

<br/>

```{r fig.cap="source: Kelleher et al., 2019", echo=FALSE, out.width="400px", fig.align="center"}
knitr::include_graphics("figures/fig_1_b_wo_caption.png")
```

???

Let's look at the following scenario an example of a phylogenetic tree:  
- we have gathered whole-genome sequence information of 5 individuals (*a-e*)
- we compare those 5 sequences with each other and list the differences (*orange diamonds*)
- based on those differences we can infer the history and common ancestors (*f-i*)  

Our model therefore predicts, that the earliest common ancestor of those 5 individuals is *i*.  
On the left hand side there happened changes at position 5 and 1 which led to a hypothetical individual *h* and so on.  

---

## .moon-gray[.bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[Inferring] .bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[whole-genome]] .bg-washed-green.b--dark-green.ba.bw2.br3.ph2.pv1[histories] .moon-gray[in .bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[large population datasets]]


???

There are several different forms of DNA changes.  
Here I'll show two of them:  

- mutation  
- recombination
**TODO: mention mutation and recombination, what is it and when do they occur.**  

---

## .moon-gray[.bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[Inferring] .bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[whole-genome] .bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[histories] in] .bg-lightest-blue.b--blue.ba.bw2.br3.ph2.pv1[large population datasets]

<br/>

```{r sequencing-costs, fig.cap="source: genome.gov/sequencingcosts", echo=FALSE, fig.align="center", fig.height=6, fig.width=10, message=FALSE}

sequencing_cost <-
  read_csv("data/Sequencing_Cost_Data_May2020.csv")

sequencing_cost %>%
  ggplot(aes(x = Date, y = `Cost per Genome`)) +
  geom_line(size = 1.5, colour = "#357edd") +
  scale_y_log10(labels=scales::dollar_format(big.mark = "'"), limits = c(1E1, 1E8)) +
  theme_xaringan() +
  theme(axis.text = element_text(size = 14),
        axis.title = element_text(size = 17)) +
  labs(x = NULL, y = "Cost per Human Genome")

```



???

In recent years, new methods were developed to sequence DNA much faster and cheaper.  
Note that the y-axis is shown as log scale! The decrease in price is enormous.

---

## .moon-gray[.bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[Inferring] .bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[whole-genome] .bg-near-white.b--moon-gray.ba.bw2.br3.ph2.pv1[histories] in] .bg-lightest-blue.b--blue.ba.bw2.br3.ph2.pv1[large population datasets]

<br/>

```{r data-sets, echo=FALSE, fig.align="center", fig.height=6, fig.width=10}
data_sets <-
  tribble(
    ~data_set, ~n,
    "Simons Genome Diversity Project", 277,
    "1000 Genomes Project", 2504,
    "UK Biobank", 487000
  )

data_sets %>%
  mutate(data_set = fct_reorder(data_set, desc(n))) %>%
  ggplot(aes(x = data_set, y = n)) +
  geom_segment(aes(xend = data_set, yend = 1), colour = "darkgrey", size = 2) +
  geom_point(fill = "#cdecff", colour = "#357edd", size = 6, stroke = 2, shape = 21) +
  geom_text(aes(label = scales::comma(n, big.mark = "'")), vjust = -0.7, size = 8, fontface = "bold", family = "Noto Sans") +
  scale_y_log10(labels = scales::comma_format(big.mark = "'")) +
  expand_limits(y = c(1, 1000000)) +
  coord_flip() +
  theme_xaringan() +
  theme(axis.text = element_text(size = 14),
        axis.title = element_text(size = 17)) +
  labs(x = NULL, y = "n")
```



???

A consequence of that is, that the size of available data sets grew at a high rate.  
With larger data sets, more efficient algorithms (in respect to time and storage space) are needed.  

---

# Current situation

- Many methods to infer phylogenetic trees
- A single tree however can't describe recombination processes, a network is required -> ancestral recombination graph (ARG)
- ARGs can't be applied to infer large networks, since they're computationally expensive


.pull-left[ 
.center[Tree]
```{r fig.cap="source: Adapted from Morrison, 2016", echo=FALSE, out.width="400px", fig.align="center"}
knitr::include_graphics("figures/morrison_2016_tree.svg")
```
]
--
.pull-right[
.center[Network]
```{r fig.cap="source: Adapted from Morrison, 2016", echo=FALSE, out.width="400px", fig.align="center"}
knitr::include_graphics("figures/morrison_2016_network.svg")
```
]
---
class: center, middle

## .bg-washed-red.b--light-red.ba.bw2.br3.ph2.pv1[Inferring] .bg-washed-yellow.b--yellow.ba.bw2.br3.ph2.pv1[whole-genome] .bg-washed-green.b--dark-green.ba.bw2.br3.ph2.pv1[histories] in .bg-lightest-blue.b--blue.ba.bw2.br3.ph2.pv1[large population datasets]

???

Summarize
